### DISCUSSION TITLE
Implementing a custom microphone component

### DISCUSSION THREAD

**Discussion by CollinBorns** [speechly/speechly#107] [Q&A] [answered]
Is there more information on how to implement the custom microphone component that I could be pointed to?

**Comment by SigmaDeltaCom** [comment]
Collin, is this for use with the SDK ?  Which language, and tech stack are you using? WEB, iOS ??

> Is there more information on how to implement the custom microphone component that I could be pointed to?

**Comment by aukkonen** [comment, marked as answer]
The answer indeed depends on the platform you are working on. If you can use one of our [Client libraries](https://docs.speechly.com/client-libraries/usage/), implementing a custom microphone component follows more or less the same principles, and is done with our [Client Library API](https://docs.speechly.com/client-libraries/client-api-reference/):
1. Initialise the client. (E.g. with our browser-client this is just a matter of creating an instance of the client.)
1. When the microphone component is pressed, call `startContext`. This starts to send audio to the Speechly API.
1. When the microphone component is released (or some other stopping condition is reached), calls `stopContext` to stop transmitting audio.

In practice you probably should also monitor the state of the client by registering a callback with `onClientStateChanged`. Then there is also the question on how to handle requesting microphone permissions from the user if working e.g. in a browser.

For an example of how this is done in our ready-made custom microphone component for the web, please see the sources:
https://github.com/speechly/speechly/blob/main/libraries/browser-ui/src/push-to-talk-button.svelte

**Reply by arzga** [reply]
For web you can now also use Speechly `browser-client` directly from a CDN without build tooling (in addition to Node use).

https://github.com/speechly/speechly/tree/main/libraries/browser-client#usage-with-browsers

Please serve the HTML page via a web server and call `client.initialize()` from a click handler for the browser to grant the mic permission.
